### PR TITLE
ch4/am: fix payload size checking during unpack

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description
The expected datatype size and incoming data size are available at
MPIDIG_recv_type_init time, check the truncation error then rather than
during the unpacking time.

MPIDIG_recv_copy_seg's reutrn need indicate whether this is last segment
or not. Since actual_unpack_bytes is not a good indicator, we correct
that to adjust in_data_sz with payload_sz directly instead.

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [x] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/warrnings/quick

**Comment by hzhou** [comment]
test:mpich/warnings/quick

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/custom
netmod: ch4:ofi
config: shmem

File: src/mpid/ch4/generic/am/mpidig_am_msg.h [hunk 1]
```diff
@@ -195,29 +195,22 @@ MPL_STATIC_INLINE_PREFIX int MPIDIG_recv_copy_seg(void *payload, MPI_Aint payloa
...
-            rreq->status.MPI_ERROR = MPIDIG_ERR_TRUNCATE(p->offset, p->offset + p->in_data_sz);
-            MPIR_STATUS_SET_COUNT(rreq->status, p->offset);
-            return 1;
-        } else if (payload_sz > actual_unpack_bytes) {
+        if (!rreq->status.MPI_ERROR && payload_sz > actual_unpack_bytes) {
             /* basic element size mismatch */
             rreq->status.MPI_ERROR =
                 MPIR_Err_create_code(MPI_SUCCESS, MPIR_ERR_RECOVERABLE, __func__, __LINE__,
                                      MPI_ERR_TYPE, "**dtypemismatch", 0);
-            return 1;
```
Comment (pavanbalaji) [inline src/mpid/ch4/generic/am/mpidig_am_msg.h:219]: Why is this error return removed?

Comment (hzhou) [reply src/mpid/ch4/generic/am/mpidig_am_msg.h:219]: Return "TRUE" here means all segments have been received (for the transport, irrelevant to how the data is being unpacked). Since a "under-packing" due to basic type mismatch does not mean there is no more segments coming, we shouldn't return "TRUE" here.

Comment (pavanbalaji) [reply src/mpid/ch4/generic/am/mpidig_am_msg.h:219]: That should be a separate patch then, with the corresponding explanation.

Comment (hzhou) [reply src/mpid/ch4/generic/am/mpidig_am_msg.h:219]: ok

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
Hi @pavanbalaji , I have split the commit and added one commit to move the freeing of iov array into `mpidig_am_msg.h`. Could you review it again?

File: src/mpid/ch4/generic/am/mpidig_am_msg.h [hunk 2]
```diff
@@ -106,7 +106,8 @@ MPL_STATIC_INLINE_PREFIX void MPIDIG_recv_copy(void *in_data, MPIR_Request * rre
                             MPIDIG_REQUEST(rreq, count),
                             MPIDIG_REQUEST(rreq, datatype), 0, &actual_unpack_bytes);
         if (!rreq->status.MPI_ERROR && in_data_sz > actual_unpack_bytes) {
-            /* If the receive buffer had enough space, but we still
+            /* Truncation error has been checked at MPIDIG_recv_type_init.
+             * If the receive buffer had enough space, but we still
```
Comment (pavanbalaji) [inline src/mpid/ch4/generic/am/mpidig_am_msg.h:118]: This comment belongs to the previous patch, which this actual change was made.

File: src/mpid/ch4/generic/am/mpidig_am_msg.h [hunk 3]
```diff
@@ -199,15 +200,17 @@ MPL_STATIC_INLINE_PREFIX int MPIDIG_recv_copy_seg(void *payload, MPI_Aint payloa
...
                 MPIR_Err_create_code(MPI_SUCCESS, MPIR_ERR_RECOVERABLE, __func__, __LINE__,
                                      MPI_ERR_TYPE, "**dtypemismatch", 0);
-            return 1;
-        } else if (p->in_data_sz == 0) {
+        }
+
+        p->in_data_sz -= payload_sz;
+        p->offset += payload_sz;
+
+        if (p->in_data_sz == 0) {
```
Comment (pavanbalaji) [inline src/mpid/ch4/generic/am/mpidig_am_msg.h:221]: This part doesn't belong to this patch.

File: src/mpid/ch4/generic/am/mpidig_am_msg.h [hunk 4]
```diff
@@ -50,6 +50,14 @@ MPL_STATIC_INLINE_PREFIX void MPIDIG_recv_init(int is_contig, MPI_Aint in_data_s
...
 }
 
+MPL_STATIC_INLINE_PREFIX void MPIDIG_recv_finish(MPIR_Request * rreq)
+{
+    /* Free the iov array if we allocated it */
+    if (MPIDIG_REQUEST(rreq, req->status) & MPIDIG_REQ_RCV_NON_CONTIG) {
+        MPL_free(MPIDIG_REQUEST(rreq, req->iov));
+    }
+}
+
```
Comment (pavanbalaji) [inline src/mpid/ch4/generic/am/mpidig_am_msg.h:60]: Unrelated patch.

Comment (hzhou) [reply src/mpid/ch4/generic/am/mpidig_am_msg.h]: I'll move this to a separate PR.

File: src/mpid/ch4/generic/am/mpidig_am_msg.h (hunk 2 above)
Comment (hzhou) [reply src/mpid/ch4/generic/am/mpidig_am_msg.h:110]: Will fix ... done.

File: src/mpid/ch4/generic/am/mpidig_am_msg.h (hunk 3 above)
Comment (hzhou) [reply src/mpid/ch4/generic/am/mpidig_am_msg.h:213]: Wait, this is the only part in this commit. Earlier we assume `actual_unpack_bytes` is always the same as `payload_sz` unless it is "truncation error", which we abort. This patch fixes it due to the understanding that `actual_unpack_bytes` may not correlate to actual "remaining in_data_sz" or `offset`, so we should directly adjust it by `payload_sz` and check returning status by whether "remaining in_data_sz" is 0.

Comment (pavanbalaji) [reply src/mpid/ch4/generic/am/mpidig_am_msg.h:213]: Using `payload_sz` is incorrect here.  `actual_unpack_bytes` was the correct amount of data moved to the user buffer.

Comment (raffenet) [reply src/mpid/ch4/generic/am/mpidig_am_msg.h:213]: > Using `payload_sz` is incorrect here. `actual_unpack_bytes` was the correct amount of data moved to the user buffer.

That's true, but `p->in_data_sz` is tracking bytes coming from the sender, not necessarily what is being placed into the recv buffer. Perhaps we should be careful about setting the offset value such that it doesn't cause an out-of-bounds access, though.

File: src/mpid/ch4/generic/am/mpidig_am_msg.h [hunk 5]
```diff
@@ -195,29 +195,22 @@ MPL_STATIC_INLINE_PREFIX int MPIDIG_recv_copy_seg(void *payload, MPI_Aint payloa
...
             /* basic element size mismatch */
             rreq->status.MPI_ERROR =
                 MPIR_Err_create_code(MPI_SUCCESS, MPIR_ERR_RECOVERABLE, __func__, __LINE__,
                                      MPI_ERR_TYPE, "**dtypemismatch", 0);
-            return 1;
-        } else if (p->in_data_sz == 0) {
+        }
+
+        p->in_data_sz -= payload_sz;
+        p->offset += payload_sz;
```
Comment (raffenet) [inline src/mpid/ch4/generic/am/mpidig_am_msg.h:211]:
```suggestion
        p->offset += actual_unpack_bytes;
```

Comment (hzhou) [reply src/mpid/ch4/generic/am/mpidig_am_msg.h:211]: This is actually a main point of this PR. The `actual_unpack_bytes` can be less than the `payload_sz` (size of the segment), so adding `actual_unpack_bytes` may result  in  the next segment offset mis-aligned. We can't simply abort either, because we can't leave the remaining segments hanging.

Comment (raffenet) [reply src/mpid/ch4/generic/am/mpidig_am_msg.h:211]: We need to make sure we receive all the data, but the buffer contents are going to be garbage. We should just make sure we don't go out of bounds.

Comment (hzhou) [reply src/mpid/ch4/generic/am/mpidig_am_msg.h:211]: At this branch, it is already an error case -- we have a hole at the end of this segment. The focus is to receive the rest of the segments. If we adjust the offset by `actual_unpack_bytes`, then the next segments will be copied at wrong offsets -- all shifted by the size of "hole". I guess it probably doesn't matter match since we already claimed this is an error and user shouldn't care about data anyway. I was thinking at least the reset of the segments can be copied correctly so minimizing the data corruption or at least reduce the amount data noise. 

Why do you think adjust by `payload_sz` would result in out-of-bound ?

Comment (raffenet) [reply src/mpid/ch4/generic/am/mpidig_am_msg.h:211]: My assumption is that if Typerep leaves a hole at the end of the unpack operation, then a future unpack starting at the end of the hole might not have enough space. This hinges on `in_data_sz` being equal to `dt_size` * `count`. But since this is the error case, it's probably not worth debating too vigorously. We could also run this through the AddressSanitizer to see if it flags anything.

Comment (hzhou) [reply src/mpid/ch4/generic/am/mpidig_am_msg.h:211]: `in_data_sz` and `dt_size` is being checked up-front: https://github.com/pmodels/mpich/blob/0462b595ab25dd0495b6882e7058995b21bef417/src/mpid/ch4/generic/am/mpidig_am_msg.h#L29-L35

I concede that it probably doesn't matter how we adjust `offset` in this case. Since it doesn't matter, I prefer the uniformity this version provides:
```
       p->in_data_sz -= payload_sz;
       p->offset += payload_sz;
```
-- otherwise, it leads readers to ponder on the difference between `payload_sz` and `actual_unpack_bytes` when they doesn't matter, thus more difficult to read.

Let me rebase and run the tests.

Comment (hzhou) [reply src/mpid/ch4/generic/am/mpidig_am_msg.h:211]: Note that, since we always use power of 2 as segment size (and headers observe 8-byte alignment), I don't think this error case (due to basic datatype mis-align) will ever arise. But since we added this error case, I think we'd better make it correct.

**Comment by hzhou** [comment]
test:mpich/ch4/most

File: src/mpid/ch4/generic/am/mpidig_am_msg.h (hunk 5 above)
Comment (hzhou) [reply src/mpid/ch4/generic/am/mpidig_am_msg.h:219]: Also, https://github.com/pmodels/mpich/blob/0462b595ab25dd0495b6882e7058995b21bef417/src/mpid/ch4/generic/am/mpidig_am_msg.h#L199-L202
`MPIR_Typerep_unpack` is supposed to check recv datatype type and size, so even if we erroneously give an out-of-bound offset, it should not actually write out-of-bound, right?

Comment (raffenet) [reply src/mpid/ch4/generic/am/mpidig_am_msg.h:219]: Dataloop code doesn't appear to take offset into it's error determination. The yaksa Typerep impl doesn't look to have the same error checks, but they also might be down in the library itself. We also aren't capturing errors from the Typerep APIs...

Comment (hzhou) [reply src/mpid/ch4/generic/am/mpidig_am_msg.h:219]: Both dataloop and yaksa have to seek to the offset before unpacking, right? I am assuming it has to bail out if the offset is beyond the datatype. Either that, or it will simply only take the remaining data, which will become 0 or negative for offset beyond datatype boundary, right?

Comment (raffenet) [reply src/mpid/ch4/generic/am/mpidig_am_msg.h:219]: Yes, but Typereps are repeatable, e.g. vector. There is no guarantee dataloop or yaksa are tracking the bounds of the (buffer,type,count). In fact, they might explicitly not do that for performance purposes.

Comment (hzhou) [reply src/mpid/ch4/generic/am/mpidig_am_msg.h:219]: > Yes, but Typereps are repeatable, e.g. vector. There is no guarantee dataloop or yaksa are tracking the bounds of the (buffer,type,count). In fact, they might explicitly not do that for performance purposes.

It has to calculate how many bytes beyond `offset` to copy, right? In the case of `offset` beyond bounds, the number of bytes will be negative or truncated to 0, right?

Comment (hzhou) [reply src/mpid/ch4/generic/am/mpidig_am_msg.h:219]: Anyway, that was my assumption. It worth checking out and confirm for sure. If my assumption doesn't hold -- typerep just blindly copy -- then I am afraid there are more places to place guard.

Comment (raffenet) [reply src/mpid/ch4/generic/am/mpidig_am_msg.h:219]: > > Yes, but Typereps are repeatable, e.g. vector. There is no guarantee dataloop or yaksa are tracking the bounds of the (buffer,type,count). In fact, they might explicitly not do that for performance purposes.
> 
> It has to calculate how many bytes beyond `offset` to copy, right? In the case of `offset` beyond bounds, the number of bytes will be negative or truncated to 0, right?

Copy size is passed in explicitly to Typerep_unpack. It is not modified by the offset.

**Comment by raffenet** [comment]
Tests are clean, so perhaps this is OK to merge as it is. Was there any xfail related that we should remove?

**Comment by hzhou** [comment]
> Tests are clean, so perhaps this is OK to merge as it is. Was there any xfail related that we should remove?

As I said, I don't expect the `dtypemismatched` error be triggered by our current code (due to default segment size in both shmmod and netmod), so I don't think the tests can clear your suspicion. We need manually create the situation to be sure :) 

There isn't any xfail related to this PR.

**Comment by raffenet** [comment]
> > Tests are clean, so perhaps this is OK to merge as it is. Was there any xfail related that we should remove?
> 
> As I said, I don't expect the `dtypemismatched` error be triggered by our current code (due to default segment size in both shmmod and netmod), so I don't think the tests can clear your suspicion. We need manually create the situation to be sure :)
> 
> There isn't any xfail related to this PR.

I was just wondering what brought this PR about. This also modifies the truncation check at the receiver, which we do have tests for in the testsuite.

**Comment by hzhou** [comment]
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch4/ucx

**Comment by hzhou** [comment]
@pavanbalaji Do you want to make a final review of this PR?
